activerecord/lib/active_record/railtie.rb

```diff
@@ -319,9 +319,26 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
-    initializer "active_record.set_signed_id_verifier_secret" do
-      ActiveSupport.on_load(:active_record) do
-        self.signed_id_verifier_secret ||= -> { Rails.application.key_generator.generate_key("active_record/signed_id") }
+    initializer "active_record.set_signed_id_verifier" do
+      config.after_initialize do |app|
+        ActiveSupport.on_load(:active_record) do
+          secret = self.signed_id_verifier_secret
+          secret_generator = if secret
+            ActiveRecord.deprecator.warn(<<~MSG)
+              ActiveRecord::Base.signed_id_verifier_secret is deprecated and will be removed in Rails ?.?. The secret will be derived from 'secret_key_base' instead.
+            MSG
+            secret.respond_to?(:call) ? secret : -> (_) { secret }
+          end
+          legacy_options = { secret_generator: secret_generator, digest: "SHA256", serializer: JSON, url_safe: true }.compact
+
+          if self.use_legacy_signed_id_verifier == :generate_and_verify
+            app.message_verifiers.prepend { |salt| legacy_options if salt == "active_record/signed_id" }
+          elsif self.use_legacy_signed_id_verifier == :verify
+            app.message_verifiers.rotate { |salt| legacy_options if salt == "active_record/signed_id" }
+          end
+
+          self.signed_id_verifier ||= app.message_verifier("active_record/signed_id")
+        end
```

Comment:
❓ Where is the best place to test this functionality?

Reply:
[`railties/test/application/configuration_test.rb`](https://github.com/rails/rails/blob/main/railties/test/application/configuration_test.rb) is a good place to test configuration from initializers.

Reply:
We had one to test the default value and added one test for each mode.